### DISCUSSION TITLE
Can docx or typst snippets be inserted via ipynb cell output MIME types?

### DISCUSSION THREAD

**Discussion by jkrumbiegel** [quarto-dev/quarto-cli#8516] [Q&A]
### Description

I want to render specific tables in docx and typst format. I've already found that I can insert the raw code for these via markdown cells with raw types `{=openxml}` or `{=typst}`, respectively. These markdown blocks themselves have `text/markdown` MIME type in the JSON underlying the saved .ipynb file. What I'm wondering and cannot find info on is whether I could also directly return such code in cells with specific MIME types, for example `text/typst` or `application/openxml` (I have no idea if there are actually MIME types corresponding to these formats). Is the raw markdown block my only option in those cases?

**Comment by mcanouil** [comment]
You can use `output: asis` and output the proper markup in your code cells.

**Reply by jkrumbiegel** [reply]
Right, but for `asis` I need to output text to `stdout` if I'm not mistaken. I'm looking at this from the perspective of the Julia language, where we usually overload `Base.show` for specific MIME types so that depending on the abilities of the display, the correct representation is used. But this system seems not to be applicable here because only a limited set of types are considered by quarto/pandoc. (It would actually help me if I could find a definitive list of these somewhere.) If I want some type to have a meaningful output no matter if I want to render my qmd file to docx, typst or html, then it becomes complicated to solve that by printing out code manually and marking all such cells as `asis`.

I've seen that knitr uses functions like `knitr::is_latex_output` to change representation based on quarto output format, but this still requires the users to if-else branch whenever such an object is displayed.

**Reply by mcanouil** [reply]
You can take a look at Jupyter, I know modules output the mime type and that Quarto uses it.
You could use the same principle: https://docs.jupyter.org/en/latest/reference/mimetype.html

Note that, not all mime type are supported by all formats, so in the end it's either you that define code/function to output the right thing or you use a code/function (from modules/libraries developed by others) that does that for you.

If you use Julia through R/knitr, then you can possibly leverage the display functions already defined in JuliaCall/knitr (https://hwborchers.github.io/).

**Reply by cderv** [reply]
Can you share an example of the Julia code that would output what you expect? 

> for example text/typst or application/openxml

We may be missing handling of those mimetype in Quarto. I believe we do detect `text/latex` and others, so I think we just need to handle a few more mime type. 

Having an example would allow to easily be sure we do the right thing, and add some tests. Can you provide that ? 

> I've seen that knitr uses functions like knitr::is_latex_output to change representation based on quarto output format, but this still requires the users to if-else branch whenever such an object is displayed.

Regarding **knitr**, there is no mimetype handling. Everything is considered in stdout and will be handle differently depending if the output is marked a `asis` output or not. So usually, **knitr** and its ecosystem of packages for **knitr** will directly handle the wrapping of raw content to be output as markdown output asis.

**Reply by jkrumbiegel** [reply]
I see, so the mechanism with the ipynb is completely separate from the mechanism with knitr. And if I understand you correctly, knitr is marking cells in the markdown it outputs with `asis` if it wants pandoc to splice these directly into the output format. For example, when a `gt` table is returned from a cell (no other action seems to be necessary to get such a table to render in, lets say, docx, from what I've seen online, for example in this thread https://github.com/quarto-dev/quarto-cli/issues/7151).

We're in the process of developing an alternative runner for qmd files that doesn't depend on Python and jupyter (https://github.com/PumasAI/QuartoNotebookRunner.jl/). But our intermediate representation is currently ipynb notebooks, and not directly markdown (which knitr seems to do).

> We may be missing handling of those mimetype in Quarto.

Just to clarify again, I made up those two mime types, or at least I couldn't find clear guidelines online what mime types - if any - are considered the standard for typst and docx documents.

**Reply by jkrumbiegel** [reply]
> Can you share an example of the Julia code that would output what you expect?

I don't have anything specific to show, as I'm still in the planning phase for how this feature should work in our runner. The basic question was whether quarto would pick up any MIME types in an ipynb for direct inclusion in a document, but it doesn't seem to be the case from what I'm reading here. The differences with knitr you pointed out seem to support that.

**Reply by cderv** [reply]
Thanks a lot for all those precisions!

I am pinging @cscheid here as I think he will be interested in this discussion. I believe the Quarto team can help with your project https://github.com/PumasAI/QuartoNotebookRunner.jl/ either by improving our doc or do some slight modification that would help provide broader support in Julia community

About [QuartoNotebookRunner.jl](https://github.com/PumasAI/QuartoNotebookRunner.jl/), what is the aim exactly? Can you explain more ?

**Reply by MichaelHatherly** [reply]
> About [QuartoNotebookRunner.jl](https://github.com/PumasAI/QuartoNotebookRunner.jl/), what is the aim exactly? Can you explain more ?

@cderv https://github.com/quarto-dev/quarto-cli/discussions/7989 explains in brief why we're doing this. Happy to discuss further if needed though :)

**Reply by cscheid** [reply]
Yeah, we would definitely like to help.

I think it might be better to move this discussion to email, though - do you mind reaching out at carlos.scheidegger@posit.co? Thanks.
